plugin/documents.py

```diff
@@ -545,13 +550,30 @@ def _render_code_lens(self, name: str, index: int, region: sublime.Region, comma
                 annotation = command["title"]
         else:
             annotation = "..."
-        annotation = '<div class="codelens">{}</div>'.format(annotation)
-        accent = self.view.style_for_scope("region.greenish markup.accent.codelens.lsp")["foreground"]
-        self.view.add_regions(self._code_lens_key(index), [region], "", "", 0, [annotation], accent)
+
+        show_code_lens = userprefs().show_code_lens
+        if show_code_lens == "phantom":
+            key = region.to_tuple()
+            try:
+                phantom = self._code_lens_phantom_regions[key]
+            except KeyError:
+                self._code_lens_phantom_regions[key] = phantom = CodeLensPhantom(self.view, region)
+
+            if annotation != "...":
+                phantom.add_annotation(annotation)
+
+        elif show_code_lens == "annotation":
+            annotation = '<div class="codelens">{}</div>'.format(annotation)
+            accent = self.view.style_for_scope("region.greenish markup.accent.codelens.lsp")["foreground"]
+            self.view.add_regions(self._code_lens_key(index), [region], "", "", 0, [annotation], accent)
+
+    def _show_code_lens_phantom(self) -> None:
+        phantoms = [phantom.to_phantom() for phantom in self._code_lens_phantom_regions.values() if phantom.is_visible()]
+        self._code_lens_phantom.update(phantoms)
 
     def _do_code_lenses_async(self) -> None:
         session = self.session_async("codeLensProvider")
-        if session and session.uses_plugin():
+        if session and session.uses_plugin() and userprefs().show_code_lens:
```

Comment:
This additional check can go away, it's already taken care of by `"disabled_capabilities"`